### ISSUE TITLE
[Bug]: Collaboration Cursor flickering, sending hundreds of requests per second when used in React Strict Mode

### ISSUE THREAD

**Issue by ryanjchr** [ueberdosis/tiptap#4482]
### Which packages did you experience the bug in?

react, extension-collaboration-cursor

### What Tiptap version are you using?

2.1.11

### What’s the bug you are facing?

When using the collaboration cursor extension with a Hocuspocus Provider in a React app with Strict Mode enabled, the extension will send up to a few hundred requests per second over the Websocket as long as the editor is focused in a way that makes the user's cursor visible. To other users the cursor will appear to flicker rapidly. Disabling React Strict Mode resolves the issue, but is an indicator that something is amiss. It is difficult to say whether it's an issue with the effects in TipTap's React components or hooks, or something deeper with the collaboration cursor extension's functionality or even the way it works with Hocuspocus.

The Websocket behaves normally when the collaboration extension is used _without_ the collaboration cursor extension, which suggests it is something specifically related to the cursor extension.

The issue occurs in all common browsers and is easy to replicate, I have set up a CodeSandbox microVM that is capable of demonstrating the issue, though it can be a little flaky. The repo can be cloned and run locally to demonstrate as well.

### What browser are you using?

Chrome

### Code example

https://codesandbox.io/p/sandbox/dreamy-marco-yc6y8h?file=%2Fsrc%2FApp.tsx%3A23%2C41

### What did you expect to happen?

The collaboration cursor should function as expected, without endlessly sending messages over the WebSocket or flickering, even in React Strict Mode.

### Anything to add? (optional)

I've attached two videos demonstrating the issue, one with Strict Mode on, and the other off.

https://github.com/ueberdosis/tiptap/assets/8773259/5e14ea8e-5a12-44e2-84f3-e63a14a7254a

https://github.com/ueberdosis/tiptap/assets/8773259/a0f7c418-6a2b-4ecd-9051-8e1a3ff2ea76

### Did you update your dependencies?

- [X] Yes, I’ve updated my dependencies to use the latest version of all packages.

### Are you sponsoring us?

- [ ] Yes, I’m a sponsor. 💖

**Comment by janthurau** [comment]
Thanks for providing this! I'll have a look at this tonight, as we've already heard this from a few others users. I am unable to reproduce this in Vue, so it's somehow related to React.

**Comment by CTNicholas** [comment]
I'm seeing this one too, in Next.js! Exact same code as above, it happens when I place a cursor in two editors (no changes need to be made to the document to start the loop).

Thanks for looking into it 🙂.

**Comment by yurtsiv** [comment]
The same issue appears on version `2.1.10`

**Comment by bdbch** [comment]
This is caused by https://github.com/yjs/y-prosemirror/issues/85 which we are using for the cursors.

In short:

- The `yCursorPlugin` Prosemirror plugin adds listeners that modify the editor inside the `yCursorPlugin`
- This works fine if you only use 1 provider per editor (so every editor has it's own provider)
- As soon as you have two plugin instances using the same provider (or React strictmode as it runs all hooks twice :( ) those listener and update events will trigger each other in a loop
- means Instance A will trigger an awareness change => Instance B will listen on this change and trigger a new awareness update on it's end => Instance A will again listen and vice versa

Right now I'm not sure what to do with this problem. I tried to throw out all of our custom code but even with only using `yCursorPlugin(this.options.provider.awareness)` causes this error.

If you want to avoid this issue - make sure to create a separate provider for your second editor with collaborative cursors. atm there is no easy fix.

**Comment by wssheldon** [comment]
For what it's worth, as mentioned by [this comment](https://github.com/liveblocks/liveblocks/issues/1161#issuecomment-1738642826) in https://github.com/liveblocks/liveblocks/issues/1161 the flickering goes away for me when doing a full build + run. The issue is only present when running in dev.

This is the behvaior for me on Next.js `13.5.3` and `@tiptap/extension-collaboration-cursor` `2.1.12`

**Comment by Philipinho** [comment]
> @wssheldon: The issue is only present when running in dev.

That’s because react strict mode only runs in dev mode.

**Comment by extrange** [comment]
I think the offending commit is 4fa3cbb97560e7d1aace6a4114c414d7ac71c4f1 (`2.1.0-rc10`), specifically:

https://github.com/ueberdosis/tiptap/blob/2bea9d1513052a2fce1be69be93d6491f5b8b33d/packages/react/src/EditorContent.tsx#L150-L158

Since `React.StrictMode` renders components twice, `props.editor` is not stable.

**Comment by prichodko** [comment]
@extrange is 💯 correct.

The implementation is incorrect and should rely on refs to store instance instead of `state`.  The lifecycle is handled by `force 

I've replaced their `useEditor` hook with the following and everything works.

```tsx
import { Editor, EditorOptions } from '@tiptap/react'
import { useEffect, useRef, useState } from 'react'

export const useEditor = (options: Partial<EditorOptions> = {}, deps = []) => {
  const editorRef = useRef<Editor | null>(null)
  const [_, forceUpdate] = useState({})

  useEffect(() => {
    let isMounted = true

    editorRef.current = new Editor(options)

    editorRef.current.on('transaction', () => {
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          if (isMounted) {
            forceUpdate({})
          }
        })
      })
    })

    return () => {
      isMounted = false
      editorRef.current?.destroy()
    }
  }, deps)

  useEffect(() => () => editorRef.current?.destroy(), [])

  return editorRef.current
}

```

**Comment by svenadlung** [comment]
Should be fixed in 2.2.2 :)
